cedar-drt/src/dafny_java_impl.rs

```diff
@@ -0,0 +1,446 @@
+        expected: Option<Value>,
+    ) -> bool {
+        let jstr = self.serialize_eval_request(request, entities, expr, expected.map(make_expr));
+        let response = self.thread.call_method(
+            self.def_authorizer,
+            "eval_str",
+            "(Ljava/lang/String;)Ljava/lang/String;",
+            &[jstr.into()],
+        );
+        match response {
+            Ok(v) => self.deserialize_eval_response(v),
+            Err(e) => {
+                self.thread
+                    .exception_describe()
+                    .expect("Failed to print exception information");
+                panic!("JVM Exception Occurred!: {:?}", e);
+            }
+        }
+    }
+
+    fn serialize_auth_request(
+        &self,
+        request: &ast::Request,
+        policies: &ast::PolicySet,
+        entities: &Entities,
+    ) -> JString {
+        let request: String = serde_json::to_string(&RequestForDefEngine {
+            request,
+            policies,
+            entities,
+        })
+        .expect("Failed to serialize request, policies, or entities");
+        self.thread
+            .new_string(request)
+            .expect("failed to create Java object for authorization request string")
+    }
+
+    fn deserialize_auth_response(&self, response: JValue) -> InterfaceResponse {
+        let jresponse: JString = response
+            .l()
+            .unwrap_or_else(|_| {
+                panic!(
+                    "expected isAuthorized_str to return an Object (String), but it returned {:?}",
+                    response
+                )
+            })
+            .into();
+        let response: String = self
+            .thread
+            .get_string(jresponse)
+            .expect("failed to get JavaStr")
+            .into();
+        self.thread
+            .delete_local_ref(*jresponse)
+            .expect("Deletion failed");
+        let d_response: DefinitionalAuthResponse = serde_json::from_str(&response).unwrap_or_else(|_| {
+            panic!(
+                "JSON response received from the definitional engine was the wrong format:\n{response}",
+            )
+        });
+
+        info!(
+            "{}{}",
+            JAVA_SERIALIZATION_MSG, d_response.serialization_nanos
+        );
+        info!(
+            "{}{}",
+            JAVA_DESERIALIZATION_MSG, d_response.deserialization_nanos
+        );
+        info!("{}{}", JAVA_AUTH_MSG, d_response.auth_nanos);
+
+        d_response.response
+    }
+
+    /// Ask the definitional engine whether `isAuthorized` for the given `request`,
+    /// `policies`, and `entities`
+    pub fn is_authorized(
+        &self,
+        request: &ast::Request,
+        policies: &ast::PolicySet,
+        entities: &Entities,
+    ) -> InterfaceResponse {
+        let (jstring, dur) =
+            time_function(|| self.serialize_auth_request(request, policies, entities));
+        info!("{}{}", RUST_SERIALIZATION_MSG, dur.as_nanos());
```

Comment:
is `RUST_SERIALIZATION_MSG` correct here?

is it actually helpful to define the message as a constant, or should we inline it here for clarity?

Reply:
Yes, `RUST_SERIALIZATION_MSG` is correct. It refers to the time taken to produce a JSON version of the input for the Java engine to consume. Not too important to have it as a constant, although it does help ensure consistency between its use in `is_authorized` and `validate`.

My latest commit moves these constants from `logger.rs` to `dafny_java_impl.rs` since that's the only place they're actually used. I also added some comments to explain their purpose.